File: 2023/minmaxanchors.m
function [anchors,anchor1data,anchor2data]=minmaxanchors(data)
    %Initialize arrays
    anchors=zeros(2000,2);%index of anchor points in data
    anchor1data=zeros(2000,2);%minimums
    anchor2data=zeros(2000,2);%maximums
    
    %find length of data
    datasize=size(data);
    datalength=datasize(1);
    
    %Extract minimums and maximums
    j=0;
    j1=0;
    j2=0;
    for i = 1:datalength
        if i >= 3
            if data(i, 1) - data(i - 1, 1) > 0
                if data(i - 1, 1) - data(i - 2, 1) < 0
                    %minimum located
                    j = j + 1;
                    j1=j1+1;
                    anchors(j, 1) = i - 1;
                    anchors(j, 2) = 1;
                    anchor1data(j1,1)=data(i-1,1);
                    anchor1data(j1,2)=i-1;
                elseif data(i - 1, 1) - data(i - 2, 1) == 0
                    %zero slope located
                    if i >= 4
                        if data(i - 2, 1) - data(i - 3, 1) < 0
                            %minimum located
                            j = j + 1;
                            j1=j1+1;
                            anchors(j, 1) = i - 1;
                            anchors(j, 2) = 1;
                            anchor1data(j1,1)=data(i-1,1);
                            anchor1data(j1,2)=i-1;
                        end
                    end
                end
            elseif data(i, 1) - data(i - 1, 1) < 0
                if data(i - 1, 1) - data(i - 2, 1) > 0
                    %maximum located
                    j = j + 1;
                    j2=j2+1;
                    anchors(j, 1) = i - 1;
                    anchors(j, 2) = 2;
                    anchor2data(j2,1)=data(i-1,1);
                    anchor2data(j2,2)=i-1;
                elseif data(i - 1, 1) - data(i - 2, 1) == 0
                    %zero slope located
                    if i >= 4
                        if data(i - 2, 1) - data(i - 3, 1) > 0
                            %maximum located
                            j = j + 1;
                            j2=j2+1;
                            anchors(j, 1) = i - 1;
                            anchors(j, 2) = 2;
                            anchor2data(j2,1)=data(i-1,1);
                            anchor2data(j2,2)=i-1;
                        end
                    end
                end
            end
        end
    end

    %reduce arrays to actual length
    anchors(j+1:2000,:)=[];
    anchor1data(j1+1:2000,:)=[];
    anchor2data(j2+1:2000,:)=[];
end    

File: 2024/minmaxanchors.py
import numpy as np

def minmaxanchors(data):
  """
  Finds minimum and maximum points as anchors in a given data array.

  Args:
    data: A numpy array containing the data points.

  Returns:
    anchors: A numpy array containing anchor points (index, flag, value).
      - index: Index of the anchor point in the data array.
      - flag: 1 for minimum, 2 for maximum.
      - value: Value of the data point at the anchor point.
    anchor1data: A numpy array containing minimum points (value, index).
    anchor2data: A numpy array containing maximum points (value, index).
  """

  # Initialize arrays
  data_size = len(data)
  anchors = np.zeros((2000, 3))
  anchor1data = np.zeros((2000, 2))
  anchor2data = np.zeros((2000, 2))

  j = 0
  j1 = 0
  j2 = 0

  # Iterate through data points (starting from 3rd element)
  for i in range(2, data_size):
    prev1, prev2, curr = data[i - 2], data[i - 1], data[i]

    # Check for minimum
    if curr > prev1 and prev1 < prev2:
      anchors[j, 0] = i - 1
      anchors[j, 1] = 1  # Flag for minimum
      anchors[j, 2] = curr
      anchor1data[j1, 0] = curr
      anchor1data[j1, 1] = i - 1
      data[i - 1, 2] = j  # Mark data point as anchor
      j += 1
      j1 += 1
    elif curr == prev1 and prev1 < prev2:  # Zero slope, check previous point
      if i >= 4 and prev2 > prev1:
        anchors[j, 0] = i - 1
        anchors[j, 1] = 1
        anchors[j, 2] = curr
        anchor1data[j1, 0] = curr
        anchor1data[j1, 1] = i - 1
        data[i - 1, 2] = j
        j += 1
        j1 += 1

    # Check for maximum
    elif curr < prev1 and prev1 > prev2:
      anchors[j, 0] = i - 1
      anchors[j, 1] = 2  # Flag for maximum
      anchors[j, 2] = curr
      anchor2data[j2, 0] = curr
      anchor2data[j2, 1] = i - 1
      data[i - 1, 2] = j
      j += 1
      j2 += 1
    elif curr == prev1 and prev1 > prev2:  # Zero slope, check previous point
      if i >= 4 and prev2 < prev1:
        anchors[j, 0] = i - 1
        anchors[j, 1] = 2
        anchors[j, 2] = curr
        anchor2data[j2, 0] = curr
        anchor2data[j2, 1] = i - 1
        data[i - 1, 2] = j
        j += 1
        j2 += 1

  # Reduce arrays to actual length
  anchors = anchors[:j]
  anchor1data = anchor1data[:j1]
  anchor2data = anchor2data[:j2]

  return anchors, anchor1data, anchor2data

